### PR TITLE
feat: update Tuple, Array and Enum target types

### PR DESCRIPTION
**Pull request by QuinnLee**
## TLDR

Sorry for the large PR. Update the target to handle enums, structs and tuples from the ABI updates.

There is additional work that needs to be done around how to make TS types more compatible with Sway Types. e.g. https://github.com/FuelLabs/fuels-ts/issues/359.

Additional work needs to be done to provide a better namespace for function input and outputs.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -81,6 +81,7 @@ For Sway types, the mapping is as follows to Typescript types:
 | b256      | '0x'-prefixed-string or Uint8Array |
 | str[]     | '0x'-prefixed-string or Uint8Array |
 | struct    | Object                             |
+| enum      | Partial<Object>                    |
```
Comment (QuinnLee) [inline README.md:84]: Going to update with https://github.com/FuelLabs/fuels-ts/issues/359

File: packages/typechain-target-fuels/src/parser/abiParser.ts [hunk 2]
```diff
@@ -74,7 +76,7 @@ export interface DocumentationResult {
  */
 function parseFunctionDeclaration(
   abiPiece: RawAbiDefinition,
-  registerStruct: (struct: TupleType) => void,
+  registerStruct: (datum: StructType | EnumType | TupleType) => void,
```
Comment (luizstacio) [inline packages/typechain-target-fuels/src/parser/abiParser.ts:79]: I see that this combination repeats a lot maybe create a single type for it;
`type DatumType = StructType | EnumType | TupleType`?

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/typechain-target-fuels/src/codegen/types.ts [hunk 3]
```diff
@@ -68,16 +74,29 @@ export function generateInputType(svmType: SvmType, options: GenerateTypeOptions
...
+    case 'enum':
       if (svmType.structName && options.useStructs) {
         return `${svmType.structName}Input`;
       }
-      return generateTupleType(svmType, (svmType) =>
+      return generateEnumType(svmType, (svmType) =>
+        generateInputType(svmType, { ...options, useStructs: true })
+      );
+    case 'struct':
+      if (svmType.structName && options.useStructs) {
```
[the remark this answers is not in the document]
Comment (QuinnLee) [reply packages/typechain-target-fuels/src/codegen/types.ts:96]: it does, but we may need to rename it.  It's a flag to tell the codegen to create the `type foo =` definition for the type. It used to be just for structs but now is for other types

File: packages/typechain-target-fuels/src/parser/abiParser.ts [hunk 4]
```diff
@@ -152,19 +165,35 @@ export function parse(
...
+
   /**
    * Registers Structs used in the abi
    */
-  function registerStruct(newStruct: TupleType): void {
-    if (structs.findIndex((s) => s.structName === newStruct.structName) === -1) {
-      structs.push(newStruct);
+  function registerComplexType(datum: StructType | EnumType | TupleType): void {
+    if (
+      datum.type === 'struct' &&
```
[the remark this answers is not in the document]
Comment (QuinnLee) [reply packages/typechain-target-fuels/src/parser/abiParser.ts]: done! I had to do some casting but it's effectively equivalent

**Pull request merged**